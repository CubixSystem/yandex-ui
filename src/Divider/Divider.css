.Divider {
    display: flex;
    align-items: center;

    box-sizing: border-box;
    width: 100%;
}

.Divider-Line {
    flex: 1 1 0;

    background-color: var(--divider-fill-color);
}

.Divider-Line:first-child:not(:only-child) {
    margin-right: var(--divider-indent-all);
}

.Divider-Line:last-child:not(:only-child) {
    margin-left: var(--divider-indent-all);
}
