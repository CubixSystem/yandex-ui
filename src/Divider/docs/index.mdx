---
id: divider
path: /components/divider

title: Divider
description: Компонент для отделения одной информации от другой.

links:
 - label: source
   url: https://github.yandex-team.ru/search-interfaces/frontend/tree/master/packages/lego-components/src/Divider
 - label: startrek
   url: https://search.yandex-team.ru/stsearch?text=Divider.ts&facet.queue=ISL&facet.type=bug&facet.status=128
 - label: figma
   url: https://www.figma.com/file/CaoHFEO0w6PaE8cVaYFDoq8i/L-E-G-O?node-id=0%3A1

tabs:
 - Examples
 - Usage
 - Playground
 - Props
---
