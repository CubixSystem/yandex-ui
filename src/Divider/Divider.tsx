import React, { FC, RefObject, CSSProperties, ElementType } from 'react';
import { cn } from '@bem-react/classname';

import { DividerLine, DividerLineProps } from './Line/Divider-Line';
import './Divider.css';

export interface DividerProps extends DividerLineProps {
    /**
     * Ссылка на корневой DOM-элемент компонента
     */
    innerRef?: RefObject<HTMLElement>;

    /**
     * Пользовательские стили
     */
    style?: CSSProperties;

    /**
     * Дополнительный класс
     */
    className?: string;

    /**
     * Компонент для рендера разделителя
     * @default 'div'
     */
    as?: ElementType;
}

export const cnDivider = cn('Divider');

/**
 * Компонент для отделения одной информации от другой.
 * @param {IDividerProps} props
 */
export const Divider: FC<DividerProps> = ({
    size = 1,
    color,
    innerRef,
    style,
    className,
    children,
    as: Component = 'div',
    ...props
}) => (
    <Component {...props} ref={innerRef} className={cnDivider(null, [className])} style={style}>
        <DividerLine size={size} color={color} />
        {children}
        {children && <DividerLine size={size} color={color} />}
    </Component>
);

Divider.displayName = cnDivider();
