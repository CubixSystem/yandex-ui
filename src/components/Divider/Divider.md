# Divider

<a href='https://github.yandex-team.ru/search-interfaces/frontend/tree/master/packages/lego-components/src/components/Divider' target='_blank'><img src='https://badger.yandex-team.ru/custom/[Исходники]/[Github
][green]/badge.svg' /></a> <a href='https://search.yandex-team.ru/stsearch?text=Divider.ts&facet.queue=ISL&facet.type=bug&facet.status=128' target='_blank'><img src='https://badger.yandex-team.ru/custom/[Известные проблемы]/[Startrek][blue]/badge.svg' /></a>

<!-- description:start -->
Компонент для отделения одной информации от другой.
<!-- description:end -->

## Пример использования

Конфигурация темы на уровне проекта:

```ts
// src/lib/theme.ts
import { configureRootTheme } from '@yandex-lego/components/Theme'
import { theme } from '@yandex-lego/components/Theme/presets/default'

configureRootTheme({ theme })
```

Использование компонента:

```ts
// src/App.ts
import React from 'react'
import { Divider } from '@yandex-lego/components/Divider'

const App = () => {
  return (
    <>
      Content
      <Divier style={{ margin: '8px 0' }} />
      Content
    </>
  )
}
```

## Примеры

### Простой разделитель

{{%story::desktop:content-divider-desktop--simple%}}

### Высота разделителя

Чтобы изменить размер разделителя, установите свойство `size` в необходимое значение.

{{%story::desktop:content-divider-desktop--size%}}

### Цвет разделителя

Чтобы изменить цвет разделителя, установите свойство `color` в необходимое значение.

{{%story::desktop:content-divider-desktop--color%}}

### Разделитель с текстом

{{%story::desktop:content-divider-desktop--with-text%}}

### Разделитель с пользовательскими стилями

Чтобы изменить внешние или внутренние отступы, установите свойство `style` с необходимыми параметрами.

{{%story::desktop:content-divider-desktop--with-styles%}}

## Свойства

<!-- props:start -->
| Свойство   | Тип                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        | По умолчанию | Описание                                  |
| ---------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ------------ | ----------------------------------------- |
| innerRef?  | `RefObject<HTMLDivElement>`                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                | —            | Ссылка на корневой DOM-элемент компонента |
| size?      | `number`                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   | `1`          | Высота разделителя в пикселях              |
| style?     | `CSSProperties`                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            | —            | Пользовательские стили                    |
| className? | `string`                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   | —            | Дополнительный класс                      |
| color?     | `string`                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   | —            | Цвет разделителя                          |
| as?        | `"symbol" \| "object" \| "a" \| "abbr" \| "address" \| "area" \| "article" \| "aside" \| "audio" \| "b" \| "base" \| "bdi" \| "bdo" \| "big" \| "blockquote" \| "body" \| "br" \| "button" \| "canvas" \| "caption" \| "cite" \| "code" \| "col" \| "colgroup" \| "data" \| "datalist" \| "dd" \| "del" \| "details" \| "dfn" \| "dialog" \| "div" \| "dl" \| "dt" \| "em" \| "embed" \| "fieldset" \| "figcaption" \| "figure" \| "footer" \| "form" \| "h1" \| "h2" \| "h3" \| "h4" \| "h5" \| "h6" \| "head" \| "header" \| "hgroup" \| "hr" \| "html" \| "i" \| "iframe" \| "img" \| "input" \| "ins" \| "kbd" \| "keygen" \| "label" \| "legend" \| "li" \| "link" \| "main" \| "map" \| "mark" \| "menu" \| "menuitem" \| "meta" \| "meter" \| "nav" \| "noindex" \| "noscript" \| "ol" \| "optgroup" \| "option" \| "output" \| "p" \| "param" \| "picture" \| "pre" \| "progress" \| "q" \| "rp" \| "rt" \| "ruby" \| "s" \| "samp" \| "script" \| "section" \| "select" \| "small" \| "source" \| "span" \| "strong" \| "style" \| "sub" \| "summary" \| "sup" \| "table" \| "tbody" \| "td" \| "textarea" \| "tfoot" \| "th" \| "thead" \| "time" \| "title" \| "tr" \| "track" \| "u" \| "ul" \| "var" \| "video" \| "wbr" \| "webview" \| "svg" \| "animate" \| "animateMotion" \| "animateTransform" \| "circle" \| "clipPath" \| "defs" \| "desc" \| "ellipse" \| "feBlend" \| "feColorMatrix" \| "feComponentTransfer" \| "feComposite" \| "feConvolveMatrix" \| "feDiffuseLighting" \| "feDisplacementMap" \| "feDistantLight" \| "feDropShadow" \| "feFlood" \| "feFuncA" \| "feFuncB" \| "feFuncG" \| "feFuncR" \| "feGaussianBlur" \| "feImage" \| "feMerge" \| "feMergeNode" \| "feMorphology" \| "feOffset" \| "fePointLight" \| "feSpecularLighting" \| "feSpotLight" \| "feTile" \| "feTurbulence" \| "filter" \| "foreignObject" \| "g" \| "image" \| "line" \| "linearGradient" \| "marker" \| "mask" \| "metadata" \| "mpath" \| "path" \| "pattern" \| "polygon" \| "polyline" \| "radialGradient" \| "rect" \| "stop" \| "switch" \| "text" \| "textPath" \| "tspan" \| "use" \| "view" \| ComponentClass<any, any> \| FunctionComponent<any>` | `'div'`      | Компонент для рендера разделителя         |
<!-- props:end -->
