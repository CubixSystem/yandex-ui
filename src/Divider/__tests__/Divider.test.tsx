import React, { createRef } from 'react';

import { createClientRender, screen } from '../../internal/testing';
import { Divider } from '../Divider';

describe('Divider', () => {
    const render = createClientRender();

    test('должен вернуть полный шаблон компонента (snapshot)', () => {
        const { container } = render(<Divider className="mix" />);

        expect(container.firstChild).toMatchSnapshot();
    });

    test('должен быть установлен корректный displayName', () => {
        expect(Divider.displayName).toBe('Divider');
    });

    test('должен устанавливать ссылку на корневой DOM элемент', () => {
        const innerRef = createRef<HTMLDivElement>();
        render(<Divider innerRef={innerRef} />);

        expect(innerRef.current).not.toBe(null);
    });

    test('должен устанавливать и изменять size через props', () => {
        const { setProps } = render(
            <Divider size={2} data-testid="element">
                content
            </Divider>,
        );

        expect(screen.getByTestId('element').firstChild).toHaveStyle({ height: '2px' });
        expect(screen.getByTestId('element').lastChild).toHaveStyle({ height: '2px' });

        setProps({ size: 10 });

        expect(screen.getByTestId('element').firstChild).toHaveStyle({ height: '10px' });
        expect(screen.getByTestId('element').lastChild).toHaveStyle({ height: '10px' });
    });

    test('должен устанавливать и изменять color через props', () => {
        render(
            <Divider color="#fc0" data-testid="element">
                content
            </Divider>,
        );

        expect(screen.getByTestId('element').firstChild).toHaveStyle({ backgroundColor: 'rgb(255, 204, 0)' });
        expect(screen.getByTestId('element').lastChild).toHaveStyle({ backgroundColor: 'rgb(255, 204, 0)' });
    });

    test('должен изменять тег компонента через props as', () => {
        render(<Divider as="span" data-testid="element" />);

        expect(screen.getByTestId('element').tagName).toBe('SPAN');
    });
});
