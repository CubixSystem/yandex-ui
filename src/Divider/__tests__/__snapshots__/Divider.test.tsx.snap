// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Divider должен вернуть полный шаблон компонента (snapshot) 1`] = `
<div
  class="Divider mix"
>
  <span
    class="Divider-Line"
    style="height: 1px;"
  />
</div>
`;
